design: Add a design doc for exposing source keys to dataflow

quodlibetor opened the pull request:
[Rendered](https://github.com/quodlibetor/materialize/blob/design-kafka-keys-in-sql/doc/developer/design/20210505_kafka_keys_in_sql.md)

Kafka is unique in our supported sources (although not all possible sources,
e.g. [DynamoDB Streams][ddb]) in that it partitions messages into a Key part and
a Value part. The Key is intended to be the equivalent of a primary key in a
database, and it is commonly used that way by Kafka users.

An important aspect of the Key is that it is a separate data section -- it never
shares backing storage with the Value.

Materialize does not support accessing the Key part of messages from our SQL
layer, which has been mentioned as a pain point on several occasions. This is a
design to resolve that pain point.

cirego commented on doc/developer/design/20210505_kafka_keys_in_sql.md:
```diff
@@ -0,0 +1,165 @@
+
+Kafka is unique in our supported sources (although not all possible sources,
+e.g. [DynamoDB Streams][ddb]) in that it partitions messages into a Key part and
+a Value part. The Key is intended to be the equivalent of a primary key in a
```
Nit: there are also timestamp and headers fields on each message.

quodlibetor replied:
I intended those to come in as the more general "metadata" future work section. I'll rephrase this a bit.

cirego commented on doc/developer/design/20210505_kafka_keys_in_sql.md:
````diff
@@ -0,0 +1,165 @@
+  ```
+
+If `INCLUDE KEY` is specified, the source schema will be augmented to prepend
+all of the values of the key as columns in the dataflow. Format types that do
````
Can you include an example query that would demonstrate how the view part would look?

quodlibetor replied:
done

cirego commented on doc/developer/design/20210505_kafka_keys_in_sql.md:
```diff
@@ -0,0 +1,165 @@
+Alternatively, if the `INCLUDE KEY AS RECORD <name>` syntax is used all fields
+will be included as fields on a record with the specified name.
+
+If any key column names colide with value columns using the bare `INCLUDE KEY`
```
I think conflict on column names will occur quite frequently, so it might be nice if we could find a way to avoid the `AS RECORD`?

For example, debezium transaction messages have an "id" field in both the key and value.

quodlibetor replied:
It generally won't make sense to use this with Debezium, or any other format that duplicates data from the value into the key, specifically because the data in the key is always already present in the value.

cirego replied:
Oh, yeah, that makes sense. I'll go ahead and resolve this then!

frankmcsherry commented on doc/developer/design/20210505_kafka_keys_in_sql.md:
```diff
@@ -0,0 +1,165 @@
+RECORD` is specified, only the projection mapping and names available to SQL
+will differ.
+
+## Alternatives
```
It seems like another alternative (unless I missed it) is "always include the key" and then rely on the *use* (or not) of that column to drive the instantiation of the key. Is it important for the source declaration to decide whether or not the key will be present in the source?

quodlibetor replied:
That is a legitimate alternative! I'll include it. The largest downsides that I see to it are:

* `SELECT *`-type queries will change their semantics, which is a compatibility-impacting change (especially undesirable for things like debezium, where the key is afaik always a duplicated subset of the value fields)
* More generally, I think that for this to be an unambiguous API we would require that key and value were always presented as the `key`/`value` record syntax, which would be both more annoying and an even larger backcompat change.

frankmcsherry replied:
My sense is (maybe wrong) that internally we'll probably end up doing something like this, and the `INCLUDE KEY` stuff will desugar to the `map` statements that put the fields in the right place. At least, that seems to be the source orthogonality direction. And with that in mind, I have no strong opinion on how the user should be obliged to interact with it!

quodlibetor replied:
I think that what you said is correct, if users spell out `INCLUDE KEY` but then don't actually use it we'll still elide it.

cirego commented on doc/developer/design/20210505_kafka_keys_in_sql.md:
```diff
@@ -0,0 +1,165 @@
+-->
+
+We could make the `AS RECORD <name>` syntax required. The proposed design is
+based on the assumption that the common case is that key fields do not conflict
```
My expectation is that column name conflicts will be quite frequent, as noted above.

quodlibetor replied:
right, let's keep the conversation on that part.

cirego reviewed (approved):
I'm very much in favor of this!

quodlibetor commented:
I've updated the proposed syntax/semantics to not always nest things in a record if the upstream type is not already a record (only relevant for formats text/bytes), and I also added an open question to the end about if we should default to using the name `key` for the same formats (text/bytes).

umanwizard commented on doc/developer/design/20210505_kafka_keys_in_sql.md:
````diff
@@ -0,0 +1,302 @@
+Enhance the Kafka create source syntax like so:
+
+```sql
+CREATE SOURCE <source-name> FROM KAFKA BROKER '<>'
+<format>
+[INCLUDE KEY [AS <key-column-name>]]
````
I think the `INCLUDE KEY` should be syntactically part of the "Kafka" section, since it is a property of Kafka brokers specifically.

quodlibetor replied:
So do you mean move the `INCLUDE KEY` to before the `<format>` specification? like so:

```sql
CREATE SOURCE <source-name> FROM KAFKA BROKER '<>'
[INCLUDE KEY [AS <key-column-name>]]
<format>
```

Technically the key depends on the combination of the fact that it's Kafka (or a future key-supporting source type) *and* the format defining how to handle Keys. Conceptually I think that this change seems fine, but I want to think on it for a bit and write out some examples.

umanwizard commented on doc/developer/design/20210505_kafka_keys_in_sql.md:
````diff
@@ -0,0 +1,302 @@
+
+```sql
+CREATE SOURCE avro_avro FROM KAFKA BROKER '...' TOPIC '...'
+KEY FORMAT AVRO USING SCHEMA '{"type": "record", "name": "boring", "fields": [
````
Did you intentionally not write `INCLUDE KEY` in this example?

quodlibetor replied:
nope, don't know how that happened.

umanwizard reviewed (approved):
Seems fine overall. Thanks for putting this together!
